ext/etc/etc.c

```diff
@@ -482,6 +482,47 @@ etc_getgrnam(VALUE obj, VALUE nam)
 #endif
 }
 
+static VALUE
+etc_getgrouplist(VALUE obj, VALUE user)
+{
+#ifdef HAVE_GETGROUPLIST
+    int ngroups = 0;
+    gid_t *groups = NULL;
+    struct group *gr;
+    struct passwd *pw;
+    const char *p = StringValueCStr(user);
+    VALUE groups_ary;
+
+    pw = getpwnam(p);
+    if (pw == NULL) rb_raise(rb_eArgError, "can't find user for %"PRIsVALUE, user);
+
+    // call first to get number of groups
+    if (getgrouplist(p, pw->pw_gid, groups, &ngroups) == -1) {
+        groups = ruby_xmalloc(ngroups * sizeof (gid_t));
+        if (groups == NULL) rb_raise(rb_eArgError, "malloc");
```

Comment:
```suggestion
        groups = ALLOC_N(gid_t, ngroups);
```

`ALLOC_N` has no potential integer overflow.
And `ruby_xmalloc` never return `NULL` but raises `NoMemoryError`.